File: AllSolutions/ArraysAndStrings/roman-to-integer_2.java
//13. Roman to Integer
//https://leetcode.com/problems/roman-to-integer/solution/
class Solution {
    public int romanToInt(String s) {
        //step: iterate from rigth to left 
        //step: sum up , roman to value 
        //      if left element is less then right element, then subtract 
        //     otherwise, sum up 
        
        //Time(n) : O(n) -> O(1) since, roman numbers are finite 
        // space(n) : O(1) 
        
        //constraints: 1 <= s.len <= 15
        //add last digit 
        int sum = getRomanToNum(s.charAt(s.length() - 1));
        //iterate from 2nd digit if any
        for(int i = s.length() - 2; i >= 0; i--) {
            int next = getRomanToNum(s.charAt(i + 1));
            int current =  getRomanToNum(s.charAt(i));
            if(current < next) {
                sum -= current;
            } else {
                sum += current;
            }
        }
        
        return sum;
    }
       
    private int getRomanToNum(char roman) {
        switch(roman) {
            case 'I': return 1;
            case 'V': return 5; 
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
        }
        return 0;
    }
}

File: AllSolutions/ArraysAndStrings/roman-to-integer_2.kt
//13. Roman to Integer
//https://leetcode.com/problems/roman-to-integer/

class Solution {
    fun romanToInt(s: String): Int {
        
        var sum = getRomanToNum(s[s.lastIndex])
     
        for(i in (s.length - 2) downTo 0 step 1) {
            val current = getRomanToNum(s[i])
            val next = getRomanToNum(s[i+1])
            if(current < next) 
                sum -= current
            else 
                sum += current
        }
        return sum
    }
    
    fun getRomanToNum(roman: Char) = when(roman) {
        'I' ->  1
        'V' ->  5
        'X' ->  10
        'L' ->  50
        'C' ->  100
        'D' ->  500
        'M' -> 1000
        else -> 0
    }
    
}
